substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrChunkWriter.java

```diff
@@ -371,43 +398,52 @@ private static int makePaddedInt(long sizeWritten) {
     }
 
     private class JfrCloseFileOperation extends JavaVMOperation {
-        private final byte[] metadataDescriptor;
-        private final JfrRepository[] repositories;
 
-        protected JfrCloseFileOperation(byte[] metadataDescriptor, JfrRepository[] repositories) {
+        protected JfrCloseFileOperation() {
             // Some of the JDK code that deals with files uses Java synchronization. So, we need to
             // allow Java synchronization for this VM operation.
             super("JFR close file", SystemEffect.SAFEPOINT, true);
-            this.metadataDescriptor = metadataDescriptor;
-            this.repositories = repositories;
         }
 
         @Override
         protected void operate() {
             changeEpoch();
-            try {
-                long constantPoolPosition = writeCheckpointEvent(repositories);
-                long metadataPosition = writeMetadataEvent(metadataDescriptor);
-                patchFileHeader(constantPoolPosition, metadataPosition);
-                file.close();
-            } catch (IOException e) {
-                Logger.log(JFR_SYSTEM, ERROR, "Error while writing file " + filename + ": " + e.getMessage());
-            }
-
-            filename = null;
-            file = null;
         }
 
+        /**
+         * We need to ensure that all JFR events that are triggered by the current thread
+         * are recorded for the next epoch. Otherwise, those JFR events could pollute the data
+         * that we currently try to persist. To ensure that, we must execute the following steps
+         * uninterruptibly:
+         * - Flush all buffers (native, Java and global) to disk
+         * - Set all Java EventWriter.notified values
+         * - Change the epoch
+         * @return Whether to notify of file changes
+         */
         @Uninterruptible(reason = "Prevent pollution of the current thread's thread local JFR buffer.")
         private void changeEpoch() {
-            // TODO: We need to ensure that all JFR events that are triggered by the current thread
-            // are recorded for the next epoch. Otherwise, those JFR events could pollute the data
-            // that we currently try to persist. To ensure that, we must execute the following steps
-            // uninterruptibly:
-            //
-            // - Flush all thread-local buffers (native & Java) to global JFR memory.
-            // - Set all Java EventWriter.notified values
-            // - Change the epoch.
+            for (IsolateThread thread = VMThreads.firstThread(); thread.isNonNull(); thread = VMThreads.nextThread(thread)) {
```

Comment:
Please add a comment that it is on purpose that thread-local buffers are not reinitialized.